### PR TITLE
W3C library implementation

### PR DESCRIPTION
**Pull request by ilya-korotya**
I used a library instead of an interface for W3C, because we already have some of these structures, but with different fields ([see](https://github.com/iden3/contracts/blob/8ab48475f3e3119df3cf7d5bae5701095cf5b6b0/contracts/interfaces/IOnchainCredentialStatusResolver.sol#L46)). In Solidity, there's no way to use two interfaces that declare the same names for structures, variables, etc.

### CODE REVIEW

File: contracts/lib/W3CLib.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        uint64 revocationNonce;
+    }
+
+    struct CredentialSchema {
+        string id;
+        string _type;
+    }
+
+    struct State {
+        string rootOfRoots;
```
Comment (OBrezhniev) [inline contracts/lib/W3CLib.sol:26]: I think it's better to have it as uint256

File: contracts/lib/W3CLib.sol [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    struct CredentialStatus {
+        string id;
+        string _type;
+        uint64 revocationNonce;
+    }
+
+    struct CredentialSchema {
+        string id;
```
Comment (OBrezhniev) [inline contracts/lib/W3CLib.sol:21]: uint256?

File: contracts/lib/W3CLib.sol [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.16;
+
+import {SmtLib} from "./SmtLib.sol";
+
+// @title a common library for W3C related data structures
+library W3CLib {
```
[the remark this answers is not in the document]
Comment (OBrezhniev) [reply contracts/lib/W3CLib.sol:7]: And W3C is organization name. Rename to VerifiableCredential (or W3CVerifiableCredential).

**Review by OBrezhniev — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IW3CVerifiableCredential.sol
Comment (AndriianChestnykh) [inline contracts/interfaces/IW3CVerifiableCredential.sol:1]: Add NatSpec to struct and functions maybe

File: contracts/interfaces/IW3CVerifiableCredential.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    struct Credential {
+        uint256 id;
+        string[] context;
+        string[] _type;
+        uint64 expirationDate;
+        uint64 issuanceDate;
+        uint256 issuer;
+        SubjectField[] credentialSubject;
```
Comment (AndriianChestnykh) [inline contracts/interfaces/IW3CVerifiableCredential.sol:53]: Curious, I didn't know it's possible that one credential goes to several subjects (array). Can you provide the link to W3C standard? This is just for my understanding.

Comment (ilya-korotya) [reply contracts/interfaces/IW3CVerifiableCredential.sol:53]: **SubjectFiled** struct describes one field in the **credentialSubject** section. To process credentialSubject from the contact in the user side, we should have: **key**, **value**, **rawValue**. For example, if you have the next credentialSubject:
```json
{
   "balance": 100,
   "name": "BookMarket"
}
```
In the solidity code we will have the next representation:
```
credentualSubject: [
{key: balance, value: 100, rawValue: 0x}
{key: name, value: poseidonHash("BookMarket"), rawValue: 0x143Adsd..
]

File: contracts/lib/PrimitiveTypeUtils.sol [hunk 5]
```diff
@@ -14,6 +14,16 @@ library PrimitiveTypeUtils {
         }
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (OBrezhniev) [reply contracts/lib/PrimitiveTypeUtils.sol]: For swapEndianness function - test if it's correct and how it performs in separate PR please, and remove comment from here.

[the remark this answers is not in the document]
Comment (ilya-korotya) [reply contracts/lib/PrimitiveTypeUtils.sol]: @OBrezhniev I removed it in the latest commit.

**Review by OBrezhniev — DISMISSED** [review]

**Pull request closed without merging**